compiler/p/runtime/PPCArrayCopy.inc

```diff
@@ -2226,6 +2236,197 @@ L.reverseQuadWordAligned_vsx_residue:
    bclr  BO_IF, CR0_EQ              ! return if no residue left.
    b     .L.reverseTableDispatch    ! table dispatch the remaining residue.
 
+
+! POWER10 offers instructions we can do arrayCopy much faster, especially with short
+! copy without branch mis-predictions
+!  1) VSX load/store with length coupled with isel allows us to do copy almost without 
+!  branches; 2) D-form VSX load/store makes it easy to do 64-byte copy loop; 3) the
+!  two sets of D-form VSX load/store are very likely fused to do 32-byte load/store
+!  at a time;
+
+! __postP10ForwardCopy:
+!  kills: r7, r8, r9, r11, cr0
+!  kills: vs32, vs33, vs8, vs9
+!
+! __postP10GenericCopy:
+!  additionally kills: r5
+
+#ifdef AIXPPC
+   .csect   __postP10ForwardCopy{PR}, 4
+.__postP10ForwardCopy:
+   .function .__postP10ForwardCopy,startproc.__postP10ForwardCopy,16,0,(endproc.__postP10ForwardCopy-startproc.__postP10ForwardCopy)
+#elif defined(LINUXPPC64)
+   .align   4
+FUNC_LABEL(__postP10ForwardCopy):
+#elif defined(LINUX)
+   .align   4
+__postP10ForwardCopy:
+#else
+__postP10ForwardCopy:
+#endif
+   startproc.__postP10ForwardCopy:
+      cmpli cr0, CmpAddr, r7, 32
+      bgt   cr0, L.p10_Forward_bt32       ! no taken branch for small copy for both fwd and reverse
+L.p10_le32:
+      sldi   r11, r7, 56
+      LXVL(vs32, r8, r11)
+      addic. r7, r7, -16
+      ISELLT(r7, 0, r7, cr0)            ! r0 here really means value 0
+      sldi   r7, r7, 56
+      addi   r8, r8, 16
+      LXVL(vs33, r8, r7)
+      STXVL(vs32, r9, r11)
+      addi   r9, r9, 16
+      STXVL(vs33, r9, r7)
+      blr
+
+L.p10_Forward_bt32:
```

Comment:
Could we rename this label to L.p10_Forward_gt32  since it seems to handle sizes > 32 bytes.

Reply:
Done